src/bundle/Controller/LocationController.php

```diff
@@ -212,6 +229,83 @@ public function copyAction(Request $request): Response
         ]));
     }
 
+    /**
+     * @param Request $request
+     *
+     * @return Response
+     *
+     * @throws InvalidArgumentException
+     */
+    public function copySubtreeAction(Request $request): Response
+    {
+        $form = $this->formFactory->copyLocation(
+            new LocationCopyData(),
+            'location_copy_subtree'
+        );
+        $form->handleRequest($request);
+
+        $location = $form->getData()->getLocation();
+
+        if ($form->isSubmitted()) {
+            $result = $this->submitHandler->handle($form, function (LocationCopyData $data) {
+                $location = $data->getLocation();
+                $newParentLocation = $data->getNewParentLocation();
+
+                $isContainer = new IsContainer($this->contentTypeService);
+
+                if (!$isContainer->isSatisfiedBy($newParentLocation)) {
+                    throw new InvalidArgumentException(
+                        '$newParentLocation',
+                        'Cannot copy location to a parent that is not a container'
+                    );
+                }
+
+                if (!$isContainer->isSatisfiedBy($location)) {
+                    throw new InvalidArgumentException(
```

Comment:
Validation should be on the form itself in order to redirect the user back to the page he came from instead of throwing generic exceptions.